src/profile-logic/transforms.js

```diff
@@ -688,6 +730,78 @@ export function collapseResource(
   });
 }
 
+export function collapseDirectRecursion(
+  thread: Thread,
+  funcToCollapse: IndexIntoFuncTable,
+  implementation: ImplementationFilter
+): Thread {
+  const { stackTable, frameTable, samples } = thread;
+  const oldStackToNewStack: Map<
+    IndexIntoStackTable | null,
+    IndexIntoStackTable | null
+  > = new Map();
+  oldStackToNewStack.set(null, null);
+  const nonRecursiveStacks = new Set([null]);
+  const newStackTable = {
+    length: 0,
+    prefix: [],
+    frame: [],
+  };
+  const funcMatchesImplementation = FUNC_MATCHES[implementation];
+
+  for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+    const prefix = stackTable.prefix[stackIndex];
+    const frameIndex = stackTable.frame[stackIndex];
+    const funcIndex = frameTable.func[frameIndex];
+
+    if (
+      // The previous stacks were collapsed or matched the funcToCollapse, check to see
+      // if this is a candidate for collapsing as well.
+      !nonRecursiveStacks.has(prefix) &&
+      // Either the function must match, or the implementation must be different.
+      (funcToCollapse === funcIndex ||
+        !funcMatchesImplementation(thread, funcIndex))
+    ) {
+      // Out of N consecutive stacks that match the function to collapse, only remove
+      // stacks that are N > 1.
+      const newPrefixStackIndex = oldStackToNewStack.get(prefix);
+      if (newPrefixStackIndex === undefined) {
+        throw new Error('newPrefixStackIndex cannot be undefined');
+      }
+      oldStackToNewStack.set(stackIndex, newPrefixStackIndex);
+    } else {
+      // Add a stack in two cases:
+      //   1. It doesn't match the collapse requirements.
+      //   2. It is the first instance of a stack to collapse, re-use the stack and frame
+      //      information for the collapsed stack.
+      const newStackIndex = newStackTable.length++;
+      const newStackPrefix = oldStackToNewStack.get(prefix);
+      newStackTable.prefix[newStackIndex] =
+        newStackPrefix === undefined ? null : newStackPrefix;
```

Comment:
`newStackPrefix` being undefined is another case that can't happen, isn't it? Because `prefix < stackIndex` and we've added all values between zero and `stackIndex` to `oldStackToNewStack` already. I think it would be good to have another `throw` here, just to make it obvious to the reader that this should not be possible.